File: src/test/java/no/fintlabs/validation/valueparsability/validators/ValueConvertingParsabilityValidatorTest.java
package no.fintlabs.validation.valueparsability.validators;

import no.fintlabs.model.configuration.entities.ValueMapping;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ValueConvertingParsabilityValidatorTest {

    ValueConvertingParsabilityValidator valueConvertingParsabilityValidator;

    @BeforeEach
    void setup() {
        valueConvertingParsabilityValidator = new ValueConvertingParsabilityValidator();
    }

    @Test
    void getTypeToValidate() {
        assertEquals(ValueMapping.Type.VALUE_CONVERTING, valueConvertingParsabilityValidator.getTypeToValidate());
    }

    @Test
    void shouldReturnTrueForStringWithValueConvertingAndInstanceFieldReferenceOfCorrectSyntax() {
        boolean valid = valueConvertingParsabilityValidator.isValid("$vc{0}$if{fornavn}");

        assertTrue(valid);
    }

    @Test
    void shouldReturnTrueForStringWithValueConvertingAndInstanceCollectionFieldReferenceOfCorrectSyntax() {
        boolean valid = valueConvertingParsabilityValidator.isValid("$vc{0}$icf{0}{fornavn}");

        assertTrue(valid);
    }

    @Test
    void shouldReturnFalseForEmptyString() {
        boolean valid = valueConvertingParsabilityValidator.isValid("");

        assertFalse(valid);
    }

    @Test
    void shouldReturnFalseForBlankString() {
        boolean valid = valueConvertingParsabilityValidator.isValid(" ");

        assertFalse(valid);
    }

    @Test
    void shouldReturnFalseForStringWithoutDynamicValues() {
        boolean valid = valueConvertingParsabilityValidator.isValid("Søknad VGS");

        assertFalse(valid);
    }

    @Test
    void shouldReturnFalseForStringWithValueConvertingWithWrongSyntax() {
        boolean valid = valueConvertingParsabilityValidator.isValid("asd$if{fornavn}");

        assertFalse(valid);
    }

    @Test
    void shouldReturnFalseForStringWithValueConvertingWithStringId() {
        boolean valid = valueConvertingParsabilityValidator.isValid("$vc(asd)$if{fornavn}");

        assertFalse(valid);
    }

    @Test
    void shouldReturnFalseForStringWithValueConverterReferenceLast() {
        boolean valid = valueConvertingParsabilityValidator.isValid("$if{fornavn}$vc(asd)");

        assertFalse(valid);
    }

    @Test
    void shouldReturnFalseForStringWithoutValueConverterReference() {
        boolean valid = valueConvertingParsabilityValidator.isValid("$if{fornavn}");

        assertFalse(valid);
    }

    @Test
    void shouldReturnFalseForStringWithoutInstanceFieldReference() {
        boolean valid = valueConvertingParsabilityValidator.isValid("$if{fornavn}");

        assertFalse(valid);
    }

    @Test
    void shouldReturnFalseForStringWithValueConvertingAndInstanceFieldReferenceOfCorrectSyntaxButAdditionalText() {
        boolean valid = valueConvertingParsabilityValidator.isValid("$vc(0)a$if{fornavn}");

        assertFalse(valid);
    }
}

File: src/test/kotlin/no/novari/flyt/configuration/validation/valueparsability/validators/ValueConvertingParsabilityValidatorTest.kt
package no.novari.flyt.configuration.validation.valueparsability.validators

import no.novari.flyt.configuration.model.configuration.entities.ValueMapping
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class ValueConvertingParsabilityValidatorTest {
    private lateinit var validator: ValueConvertingParsabilityValidator

    @BeforeEach
    fun setUp() {
        validator = ValueConvertingParsabilityValidator()
    }

    @Test
    fun `reports the value type it validates`() {
        assertEquals(ValueMapping.Type.VALUE_CONVERTING, validator.getTypeToValidate())
    }

    @Test
    fun `accepts a value converting reference combined with an instance field reference`() {
        assertTrue(validator.isValid("\$vc{0}\$if{fornavn}"))
    }

    @Test
    fun `accepts a value converting reference combined with an instance collection field reference`() {
        assertTrue(validator.isValid("\$vc{0}\$icf{0}{fornavn}"))
    }

    @Test
    fun `rejects an empty string`() {
        assertFalse(validator.isValid(""))
    }

    @Test
    fun `rejects a blank string`() {
        assertFalse(validator.isValid(" "))
    }

    @Test
    fun `rejects a string without dynamic values`() {
        assertFalse(validator.isValid("Søknad VGS"))
    }

    @Test
    fun `rejects a value converting reference with wrong syntax`() {
        assertFalse(validator.isValid("asd\$if{fornavn}"))
    }

    @Test
    fun `rejects a value converting reference whose id is not numeric`() {
        assertFalse(validator.isValid("\$vc(asd)\$if{fornavn}"))
    }

    @Test
    fun `rejects a string where the value converting reference comes last`() {
        assertFalse(validator.isValid("\$if{fornavn}\$vc(asd)"))
    }

    @Test
    fun `rejects a string without a value converting reference`() {
        assertFalse(validator.isValid("\$if{fornavn}"))
    }

    @Test
    fun `rejects a string without an instance field reference`() {
        assertFalse(validator.isValid("\$if{fornavn}"))
    }

    @Test
    fun `rejects an otherwise valid reference followed by additional text`() {
        assertFalse(validator.isValid("\$vc(0)a\$if{fornavn}"))
    }
}
